Fix issue 11051 - Unmatched case in a final switch should always throw

AndrejMitrovic opened the pull request:
Keep a HALT instruction in a final switch statement if the function is @safe and `-release` mode is enabled.

Let's see what the autotester has to say..

andralex commented:
@AndrejMitrovic doesn't like it :)

AndrejMitrovic commented:
What's really odd is that it only fails on cross-machine builds (e.g. 32-bit host 64-bit target and vice-versa).     Needs more work. :)

AndrejMitrovic commented:
@WalterBright I'm confused why this is failing on cross-targets (meaning `-m32` on a 64-bit system and vice-versa). The code works well when the target and the host are the same. 

What the PR does is: for `safe` functions a `default` statement is added to a `final switch` statement. The default case has a halt instruction. It's even displayed by druntime with a stack trace when it's hit.

But on cross-targets the app seems to just crash.

WalterBright commented:
> Keep a HALT instruction in a final switch statement if the function is @safe and -release mode is enabled.

This should be documented in the "what does @safe do" section of the spec.

WalterBright commented:
I don't know why it is failing. Are you able to reproduce it?

AndrejMitrovic commented:
Ok perhaps this has nothing to do with the PR itself. Here's a test-case I tried on Windows:

```D
module test;

void main()
{
    assert(0);
}
```

When building with `-release -m32` running this I get:

```
object.Error@(0): assert(0) or HLT instruction
```

But when building with `-release -m64` and running I get a crash dialog instead. Why is the behavior different whether `-m32` or `-m64` is used? Is this a bug?

AndrejMitrovic commented:
Ah I think I'm beginning to understand. On Win32 DMD uses Windows SEH so it can handle a halt, but not on other platforms.So maybe I need to rework the test-cases a little bit to make them pass.

Geod24 commented:
Wake up @AndrejMitrovic ;)

AndrejMitrovic committed:
Fix issue 11051

Keep a HALT instruction in a final switch statement if
the function is @safe and -release mode is enabled.

andrej-mitrovic commented:
Well, I've rebased it, but I think it will still be failing.

AndrejMitrovic commented:
Superseded: https://github.com/dlang/dmd/pull/14841
